### PR TITLE
go_1_21: init at 1.21rc3

### PR DESCRIPTION
**Pull request by zowoq**
###### Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes (or backporting 23.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Review by qbit — APPROVED** [review]
LGTM, and it builds fine too!

**Comment by zowoq** [comment]
Seems 1.21 breaks the `buildGoModule` FOD fetcher:
```
GOPROXY list is not the empty string, but contains no entries
```

Looks like we just need to fix `buildGoModule` in a backwards compatible way, the FOD hashes are still the same.

**Comment by ThinkChaos** [comment]
Go 1.21 will download and run any Go toolchain built code references by default. I think this is something NixOS should disable, or at minimum communicate about. I'm not even convinced this behavior will work out of the box on NixOS since the `go` downloaded toolchains will presumably be pre-build and dynamically linked (this is conjecture).  
See [Go Toolchains](https://tip.golang.org/doc/toolchain). The [Go toolchain selection](https://tip.golang.org/doc/toolchain#select) section should be enough for a quick overview.

Has this been discussed? Is there a NixOS policy that applies here?  
Has this feature been tested?

**Comment by zowoq** [comment]
Downloading a toolchain isn't something we'd want to allow, I'll disable it. Very likely that it wouldn't work anyway.

**Comment by zowoq** [comment]
@ofborg eval

File: pkgs/build-support/go/module.nix [hunk 1]
```diff
@@ -85,6 +87,7 @@ let
       runHook preConfigure
       export GOCACHE=$TMPDIR/go-cache
       export GOPATH="$TMPDIR/go"
+      export GOPROXY="''${GOPROXY:-direct}" # respect impureEnvVars
```
Comment (qbit) [inline pkgs/build-support/go/module.nix:90]: Personally I think keeping it default (and making it overridable is better. Having sumdb backed checking seems like a win to me.

Comment (zowoq) [reply pkgs/build-support/go/module.nix:90]: > Personally I think keeping it default 

Do you mean the leaving it unset or setting `"https://proxy.golang.org,direct"`?

I'd prefer not to set it at but but need to do something to address https://github.com/NixOS/nixpkgs/pull/238100#issuecomment-1601932975.

> and making it overridable is better

Besides setting it via `impureEnvVars`?

Comment (qbit) [reply pkgs/build-support/go/module.nix:90]: From my understanding of:
>  When GOTOOLCHAIN is set to local, the go command always runs the bundled Go toolchain.

We only need to set GOTOOLCHAIN to address the toolchain downloading concern.. Setting GOPROXY will impact module fetching as well as (if we didn't already have it as `local` toolchain fetching)

Comment (zowoq) [reply pkgs/build-support/go/module.nix:90]: What does GOTOOLCHAIN have to do with the GOPROXY error?

Comment (qbit) [reply pkgs/build-support/go/module.nix:90]: oh, my bad, I clicked it and read the comment below ;D

In that case, I think setting it to `https://proxy.golang.org,direct` is the best approach (assuming it fixes the error).

Comment (ThinkChaos) [reply pkgs/build-support/go/module.nix:90]: The `''` in the var value seems like a typo to me?

Comment (zowoq) [reply pkgs/build-support/go/module.nix:90]: > In that case, I think setting it to https://proxy.golang.org,direct is the best approach (assuming it fixes the error).

I expect people probably won't like that `proxy.golang.org` is set here even although it is the default that we've been using anyway.

> The `''` in the var value seems like a typo to me?

It isn't a typo, need to escape `${}`.

**Comment by colemickens** [comment]
I have a version of this I'm maintaining as well. A few diffs, I didn't have to set as many env vars as you did.

I also rewrote `print-hashes` to not have duplicates and also added a 121 bootstrap.

This means I have Go cross-compiling working for my riscv64 targets, which is nice, since they have published bootstrap binary builds for riscv64.

Please feel free to steal commits: https://github.com/colemickens/nixpkgs/commits/go121

**Comment by zowoq** [comment]
> I also rewrote `print-hashes` to not have duplicates

I don't know what this means? With `print-hashes.sh` I get the same as you have for `1.21rc3`?

> and also added a 121 bootstrap.

> This means I have Go cross-compiling working for my riscv64 targets, which is nice, since they have published bootstrap binary builds for riscv64.

Will deal with stuff like this in a follow up, primary concern with this PR is aarch64/x86_64 on darwin/linux.

**Comment by zowoq** [comment]
> Will deal with stuff like this in a follow up

I'll wait for `go_1_21` to land in master and continue in https://github.com/NixOS/nixpkgs/pull/246935.

File: pkgs/build-support/go/module.nix (hunk 1 above)
Comment (zowoq) [reply pkgs/build-support/go/module.nix]: The `GOPROXY` issue was fixed in https://github.com/NixOS/nixpkgs/commit/cbc976a97c3372e1eec5db021db994b85e098d12 so I'll revert this in https://github.com/NixOS/nixpkgs/pull/250492.

**Pull request merged**